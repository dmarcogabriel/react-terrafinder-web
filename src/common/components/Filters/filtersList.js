export const PROPERTY_KINDS = [
  {
    key: 'p1',
    value: '',
    name: 'Nenhum',
  },
  {
    key: 'p2',
    value: 'Fazenda',
    name: 'Fazenda',
  },
];

export const STATES = [
  {
    key: 'sp1',
    name: 'sp',
    value: 'sp',
  },
  {
    key: 'pr1',
    name: 'PR',
    value: 'PR',
  },
];
