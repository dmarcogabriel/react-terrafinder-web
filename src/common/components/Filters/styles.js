import styled, { css } from 'styled-components';
import Button from 'common/components/atm/Button';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  ${({ theme: { color, size, shadow } }) => css`
    background: ${color.white.df};
    box-shadow: ${shadow};
    width: ${size.full};
  `}
`;

export const SearchButton = styled(Button).attrs({ modifiers: 'primary' })`
  ${({ theme: { fontSize, size } }) => css`
    font-size: ${fontSize.md};
    padding: ${size.lg};
  `}
`;
