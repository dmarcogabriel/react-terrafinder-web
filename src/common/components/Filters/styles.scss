@import '~styles/';

.Filters__container {
  display: flex;
  flex-direction: column;
  background: $white;
  box-shadow: $shadow-base;
  width: 100%;

  .searchButton {
    background: $primary-dark;
    color: $white;
    font-size: $size-md;
    padding: $size-lg $size-lg;
  }
}
