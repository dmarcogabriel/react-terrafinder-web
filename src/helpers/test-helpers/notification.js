export const mockUseNotification = (showNotification = () => {}) => {
  const mockShowNotification = jest.fn(showNotification);

  jest.mock('hooks/useNotification', () => ({
    useNotification: () => ({
      showNotification: mockShowNotification,
    }),
    NOTIFICATION_TYPES: { SUCCESS: 'success', ERROR: 'error' },
  }));

  return { mockShowNotification };
};
