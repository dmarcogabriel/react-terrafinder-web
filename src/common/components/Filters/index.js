import React from 'react';
import { Select, Button } from 'common/components';
import RangeInput from 'common/components/RangeInput';
import { useFormik } from 'formik';
import cn from 'classnames';
import './styles.scss';

import { PROPERTY_KINDS, STATES } from './filters';

export default function Filters({ onSubmit, className }) {
  const { values, handleChange, handleSubmit } = useFormik({
    initialValues: {
      propertyKind: '',
      state: '',
      size: '',
      amount: '',
    },
    onSubmit() {
      // Here we remove empty fields
      Object.keys(values).forEach((key) => {
        if (!values[key]) delete values[key];
      });

      onSubmit(values);
    },
  });

  return (
    <div className={cn('Filters__container', className && className)}>
      <Select
        id="propertyKind"
        label="Tipo de Propriedade"
        options={PROPERTY_KINDS}
        value={values.propertyKind}
        onChange={(e) => handleChange('propertyKind')(e.name)}
      />

      <RangeInput
        label="Área do Imóvel"
        onChange={(e) => handleChange('size')(JSON.stringify(e))}
      />

      <Select
        label="Estado"
        options={STATES}
        onChange={(e) => handleChange('state')(e.name)}
      />

      <RangeInput
        label="Intervalo de Preço"
        onChange={(e) => handleChange('amount')(JSON.stringify(e))}
      />

      <Button className="searchButton" onClick={handleSubmit}>
        Procurar Imóvel
      </Button>
    </div>
  );
}
