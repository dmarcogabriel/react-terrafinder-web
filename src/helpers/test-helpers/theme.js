import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { theme } from '../../styles/themes';

export const withTheme = (component) => (
  <ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component) =>
  renderer.create(<ThemeProvider theme={theme}>{component}</ThemeProvider>);
