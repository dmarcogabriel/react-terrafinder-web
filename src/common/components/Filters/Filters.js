import React from 'react';
import { useFormik } from 'formik';
import { SelectInput, RangeInput } from 'common/components';
import { moneyFormat } from 'utils/formatters';
import { Card, Box, Button } from '@mui/material';
import { PROPERTY_KINDS, STATES } from './filtersList';

export const Filters = ({ onSubmit }) => {
  const { values, handleChange, handleSubmit } = useFormik({
    initialValues: {
      propertyKind: '',
      state: '',
      size: '[0, 9999]',
      amount: '[0, 999999999]',
    },
    onSubmit() {
      const filters = {};
      if (values.propertyKind) {
        filters.propertyKind = values.propertyKind;
      }
      if (values.state) {
        filters.state = values.state;
      }
      if (values.size) {
        filters.size = values.size;
      }
      if (values.amount) {
        filters.amount = values.amount;
      }

      onSubmit(filters);
    },
  });

  const sizeRangeLabelFormat = (value) => `${value}ah`;

  const amountRangeLabelFormat = (value) => moneyFormat(value);

  return (
    <Card data-testid="filters" sx={{ width: { xs: '80%', md: '50%' } }}>
      <Box sx={{ p: 3 }}>
        <SelectInput
          noValidation
          label="Tipo de Propriedade"
          options={PROPERTY_KINDS}
          onChange={handleChange('propertyKind')}
          dataTestId="propertyKind"
          value={values.propertyKind}
        />

        <RangeInput
          value={values.size}
          dataTestId="propertySize"
          label="Área do Imóvel"
          min={10}
          max={5000}
          minDistance={100}
          step={100}
          valueLabelFormat={sizeRangeLabelFormat}
          onChange={(e) => handleChange('size')(JSON.stringify(e))}
        />

        <SelectInput
          noValidation
          dataTestId="state"
          label="Estado"
          options={STATES}
          value={values.state}
          onChange={handleChange('state')}
        />

        <RangeInput
          value={values.amount}
          dataTestId="amount"
          min={0}
          max={99999999}
          step={1000}
          label="Intervalo de Preço"
          valueLabelFormat={amountRangeLabelFormat}
          onChange={(e) => handleChange('amount')(JSON.stringify(e))}
        />
      </Box>

      <Button
        variant="contained"
        color="warning"
        fullWidth
        sx={{ py: 2 }}
        data-testid="submitButton"
        onClick={handleSubmit}
      >
        Procurar Imóvel
      </Button>
    </Card>
  );
};
