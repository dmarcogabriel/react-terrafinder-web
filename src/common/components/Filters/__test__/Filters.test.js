import React from 'react';
import ReactDOM from 'react-dom';
import { render, fireEvent } from '@testing-library/react';
import { renderWithTheme, withTheme } from 'helpers/test-helpers/theme';
import Filters from '..';

describe('common/components/Filters', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(withTheme(<Filters />), div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should change background color', () => {
    const { getByTestId } = render(withTheme(<Filters className="bg-blue" />));

    const filters = getByTestId('filters');

    expect(filters.className).toContain('bg-blue');
  });

  it('should submit without values', () => {
    const mockFunction = jest.fn();
    const { getByTestId } = render(withTheme(<Filters className="bg-blue" />));

    const submitButton = getByTestId('submitButton');
    fireEvent.click(submitButton);

    expect(mockFunction).not.toHaveBeenCalled();
  });

  it('should change property kind value', () => {
    const { getByTestId } = render(withTheme(<Filters />));

    const propertyKind = getByTestId('propertyKind');
    fireEvent.click(propertyKind);

    const option = getByTestId('option-farm');
    fireEvent.click(option);

    const value = getByTestId('propertyKindValue');
    expect(value).toHaveTextContent('Fazenda');
  });

  it('should change property size value', () => {
    const { getByTestId } = render(withTheme(<Filters onSubmit={jest.fn()} />));

    const propertySize = getByTestId('propertySize');
    fireEvent.click(propertySize);

    const minInput = getByTestId('minInput');
    const maxInput = getByTestId('maxInput');
    const okButton = getByTestId('okButton');

    fireEvent.change(minInput, { target: { value: '50' } });
    fireEvent.change(maxInput, { target: { value: '100' } });
    fireEvent.click(okButton);

    const value = getByTestId('propertySizeValue');
    expect(value).toHaveTextContent('50 até 100');
  });

  it('should change state value', () => {
    const { getByTestId } = render(withTheme(<Filters />));

    const state = getByTestId('state');
    fireEvent.click(state);

    const option = getByTestId('option-sp1');
    fireEvent.click(option);

    const value = getByTestId('stateValue');
    expect(value).toHaveTextContent('sp');
  });

  it('should change amount value', () => {
    const { getByTestId } = render(withTheme(<Filters />));

    const amount = getByTestId('amount');
    fireEvent.click(amount);

    const minInput = getByTestId('minInput');
    const maxInput = getByTestId('maxInput');
    const okButton = getByTestId('okButton');

    fireEvent.change(minInput, { target: { value: '10000' } });
    fireEvent.change(maxInput, { target: { value: '1000000' } });
    fireEvent.click(okButton);

    const value = getByTestId('amountValue');
    expect(value).toHaveTextContent('10000 até 1000000');
  });

  it('matches snapshot', () => {
    const tree = renderWithTheme(<Filters />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
