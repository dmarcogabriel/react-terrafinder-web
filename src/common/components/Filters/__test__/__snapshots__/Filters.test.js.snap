// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/components/Filters matches snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-wlvy8x-MuiPaper-root-MuiCard-root"
  data-testid="filters"
>
  <div
    className="MuiBox-root css-19kzrtu"
  >
    <div
      className="MuiBox-root css-pcfq5i"
    >
      <div
        className="MuiBox-root css-1663l76"
      >
        <label
          className="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
        >
          Tipo de Propriedade
        </label>
      </div>
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-sizeSmall MuiSelect-root css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          className="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            className="notranslate"
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8203;",
              }
            }
          />
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          data-testid="propertyKind"
          onAnimationStart={[Function]}
          onChange={[Function]}
          tabIndex={-1}
          value=""
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-1a1ihhj"
          >
            <span
              className="notranslate"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="MuiBox-root css-0"
    >
      <label
        className="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
      >
        Área do Imóvel
      </label>
      <span
        className="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root css-187mznn-MuiSlider-root"
        data-testid="propertySize"
        onMouseDown={[Function]}
      >
        <span
          className="MuiSlider-rail css-14pt78w-MuiSlider-rail"
        />
        <span
          className="MuiSlider-track css-1gv0vcd-MuiSlider-track"
          style={
            Object {
              "left": "0%",
              "width": "100%",
            }
          }
        />
        <span
          className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
          data-index={0}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          style={
            Object {
              "left": "0%",
              "pointerEvents": "none",
            }
          }
        >
          <input
            aria-orientation="horizontal"
            aria-valuemax={5000}
            aria-valuemin={10}
            aria-valuenow={10}
            data-index={0}
            disabled={false}
            max={5000}
            min={10}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            step={100}
            style={
              Object {
                "border": 0,
                "clip": "rect(0 0 0 0)",
                "direction": "ltr",
                "height": "100%",
                "margin": -1,
                "overflow": "hidden",
                "padding": 0,
                "position": "absolute",
                "whiteSpace": "nowrap",
                "width": "100%",
              }
            }
            type="range"
            value={10}
          />
          <span
            aria-hidden={true}
            className="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
          >
            <span
              className="MuiSlider-valueLabelCircle"
            >
              <span
                className="MuiSlider-valueLabelLabel"
              >
                10ah
              </span>
            </span>
          </span>
        </span>
        <span
          className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
          data-index={1}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          style={
            Object {
              "left": "100%",
              "pointerEvents": "none",
            }
          }
        >
          <input
            aria-orientation="horizontal"
            aria-valuemax={5000}
            aria-valuemin={10}
            aria-valuenow={5000}
            data-index={1}
            disabled={false}
            max={5000}
            min={10}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            step={100}
            style={
              Object {
                "border": 0,
                "clip": "rect(0 0 0 0)",
                "direction": "ltr",
                "height": "100%",
                "margin": -1,
                "overflow": "hidden",
                "padding": 0,
                "position": "absolute",
                "whiteSpace": "nowrap",
                "width": "100%",
              }
            }
            type="range"
            value={5000}
          />
          <span
            aria-hidden={true}
            className="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
          >
            <span
              className="MuiSlider-valueLabelCircle"
            >
              <span
                className="MuiSlider-valueLabelLabel"
              >
                5000ah
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      className="MuiBox-root css-pcfq5i"
    >
      <div
        className="MuiBox-root css-1663l76"
      >
        <label
          className="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
        >
          Estado
        </label>
      </div>
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-sizeSmall MuiSelect-root css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          className="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            className="notranslate"
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8203;",
              }
            }
          />
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          data-testid="state"
          onAnimationStart={[Function]}
          onChange={[Function]}
          tabIndex={-1}
          value=""
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-1a1ihhj"
          >
            <span
              className="notranslate"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="MuiBox-root css-0"
    >
      <label
        className="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
      >
        Intervalo de Preço
      </label>
      <span
        className="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root css-187mznn-MuiSlider-root"
        data-testid="amount"
        onMouseDown={[Function]}
      >
        <span
          className="MuiSlider-rail css-14pt78w-MuiSlider-rail"
        />
        <span
          className="MuiSlider-track css-1gv0vcd-MuiSlider-track"
          style={
            Object {
              "left": "0%",
              "width": "100%",
            }
          }
        />
        <span
          className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
          data-index={0}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          style={
            Object {
              "left": "0%",
              "pointerEvents": "none",
            }
          }
        >
          <input
            aria-orientation="horizontal"
            aria-valuemax={99999999}
            aria-valuemin={0}
            aria-valuenow={0}
            data-index={0}
            disabled={false}
            max={99999999}
            min={0}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            step={1000}
            style={
              Object {
                "border": 0,
                "clip": "rect(0 0 0 0)",
                "direction": "ltr",
                "height": "100%",
                "margin": -1,
                "overflow": "hidden",
                "padding": 0,
                "position": "absolute",
                "whiteSpace": "nowrap",
                "width": "100%",
              }
            }
            type="range"
            value={0}
          />
          <span
            aria-hidden={true}
            className="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
          >
            <span
              className="MuiSlider-valueLabelCircle"
            >
              <span
                className="MuiSlider-valueLabelLabel"
              >
                R$ 0,00
              </span>
            </span>
          </span>
        </span>
        <span
          className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
          data-index={1}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          style={
            Object {
              "left": "100%",
              "pointerEvents": "none",
            }
          }
        >
          <input
            aria-orientation="horizontal"
            aria-valuemax={99999999}
            aria-valuemin={0}
            aria-valuenow={99999999}
            data-index={1}
            disabled={false}
            max={99999999}
            min={0}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            step={1000}
            style={
              Object {
                "border": 0,
                "clip": "rect(0 0 0 0)",
                "direction": "ltr",
                "height": "100%",
                "margin": -1,
                "overflow": "hidden",
                "padding": 0,
                "position": "absolute",
                "whiteSpace": "nowrap",
                "width": "100%",
              }
            }
            type="range"
            value={99999999}
          />
          <span
            aria-hidden={true}
            className="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
          >
            <span
              className="MuiSlider-valueLabelCircle"
            >
              <span
                className="MuiSlider-valueLabelLabel"
              >
                R$ 99.999.999,00
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
  </div>
  <button
    className="MuiButton-root MuiButton-contained MuiButton-containedWarning MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root css-13wdhlw-MuiButtonBase-root-MuiButton-root"
    data-testid="submitButton"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Procurar Imóvel
  </button>
</div>
`;
